### PR TITLE
DBZ-1656 Blogpost about Debezium Camel component

### PR DESCRIPTION
**Pull request by jpechane**

### CODE REVIEW

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 1]
```diff
@@ -0,0 +1,290 @@
+= Integration Scenarios with Debezium and Camel
+jpechane
+:awestruct-tags: [ camel, integration, quarkus ]
+:awestruct-layout: blog-post
+
+One of the typical Debezium uses cases is to use change data capture to integrate a legacy system with other systems in the organization.
+There are multiple ways how to achieve this goal
+
+* Write data to Kafka using Debezium and follow with a combination of Kafka Streams pipelines and Kafka Connect connectors to deliver the changes to other systems
+* Use Debezium Embedded engine in a Java standalone application and write the integration code using plain Java
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
* Use Debezium Embedded engine in a Java standalone application and write the integration code using plain Java; that's often used to send change events to alternative messaging infrastructure such as Amazon Kinesis, Google Pub/Sub etc.
```

Also please add a link to the embedded engine docs.

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+* Use existing integration framework or service bus to express the pipeline logic
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
* Use an existing integration framework or service bus to express the pipeline logic
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+This article is focusing on the third option - a dedicated integration framework.
+
+
+== Apache Camel
+
+https://camel.apache.org/[Camel] is an open-source integration framework that enables the developer to read and write data from and to disparate systems and services.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Also transform?

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+It provides a large amount of https://camel.apache.org/components/latest/[components] that either provide interfaces to 3rd party systems or provides an implementation of https://en.wikipedia.org/wiki/Enterprise_Integration_Patterns[enterprise integration patterns].
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
It provides a large amount of https://camel.apache.org/components/latest/[components] that either provide interfaces to 3rd party systems or offer an implementation of https://en.wikipedia.org/wiki/Enterprise_Integration_Patterns[enterprise integration patterns].
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+This combination allows the developer to easily connect to target systems and express integration pipelines using declarative DSL.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
This combination allows the developer to easily connect to target systems and express integration pipelines using a declarative DSL.
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+
+=== Camel and Debezium
+
+https://camel.apache.org/blog/Camel3-Whatsnew/[Camel 3] has been released by the end of the year 2019 and in addition to the major re-architecture new Debezium components has been added to the codebase.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
https://camel.apache.org/blog/Camel3-Whatsnew/[Camel 3] has been released by the end of the year 2019 and in addition to the major re-architecture new Debezium components have been added to the codebase.
```

Also there should be a sentence at first introducing the two general options:

* Using the Camel Kafka Connect sinks
* Using the DBZ Camel component (that's what we're discussing here)

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+* https://camel.apache.org/components/latest/debezium-mysql-component.html[Debezium MySQL Connector Component]
+* https://camel.apache.org/components/latest/debezium-postgres-component.html[Debezium PostgreSQL Connector Component]
+* https://camel.apache.org/components/latest/debezium-sqlserver-component.html[Debezium SQL Server Connector Component]
+* https://camel.apache.org/components/latest/debezium-mongodb-component.html[Debezium MongoDB Connector Component]
+
+As you can see every Debezium connector is represented by its dedicated component.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Not quite _every_ :) Perhaps "every non-incubating"?

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+The advantage of this solution is the complete isolation of dependencies and the type-safe configuration of connector instances.
+
+Internally the component exposes a Debezium endpoint with Camel consumer that encapsulates an instance of https://debezium.io/documentation/reference/1.1/operations/embedded.html[Debezium embedded engine].
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: What's a "Camel consumer" in this context? This sentence is a bit hard to grok. Link to EE should be above.

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+
+== Example
+
+We have prepared https://github.com/debezium/debezium-examples/tree/master/camel-component[an example] of how to deploy an integration pipeline using Debezium component.
+
+The source of change data is a simple Question & Answer (QnA) application.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
The source of change data is a simple Question & Answer (Q & A) application.
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+The changes generated by the application are passed to the Camel pipeline that sends emails via an SMTP server and posts a tweet on a provided Twitter account.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: There should be a sentence on the nature of the changes (new answer records etc.)

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+
+There are multiple components in the solution topology.
+
+.The Deployment topology
+[#img-topology]
+[caption="Figure 1: "]
+image::camel-component-topology.svg[Deployment topology]
+
+&nbsp; +
+The QaN application runs as a Quarkus application and exposes a REST API to create questions and answers.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
The Q & A application is implemented using the https://quarkus.io/[Quarkus] stack and exposes a REST API to create questions and answers.
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+The application stores the data in the PostgreSQL database.
+The Camel route runs as a plain Java application that uses https://infinispan.org/[Infinispan] to store its state and sends messages to https://github.com/mailhog/MailHog[MailHog] SMTP server running in a container and into an associated Twitter account.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Add a sentence on what's the state used for.

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+
+==== Question & Answer application
+
+The source application is a simple application developed using https://quarkus.io/[Quarkus framework].
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Link should be above upon first mention.

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+It provides two entities `Question` and `Answer` with `1:n` relation stored in the PostgreSQL database.
+
+.The QnA backend service ERD
+[#img-qna-erd]
+[caption="Figure 2: "]
+image::camel-component-erd.svg[QnA backend service ERD]
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
image::camel-component-erd.svg[Q & A backend service ERD]
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+As you can see every Debezium connector is represented by its dedicated component.
+The advantage of this solution is the complete isolation of dependencies and the type-safe configuration of connector instances.
+
+Internally the component exposes a Debezium endpoint with Camel consumer that encapsulates an instance of https://debezium.io/documentation/reference/1.1/operations/embedded.html[Debezium embedded engine].
+
+
+== Example
+
+We have prepared https://github.com/debezium/debezium-examples/tree/master/camel-component[an example] of how to deploy an integration pipeline using Debezium component.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
We have prepared https://github.com/debezium/debezium-examples/tree/master/camel-component[an example] of how to deploy an integration pipeline using the Camel Debezium component.
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+image::camel-component-erd.svg[QnA backend service ERD]
+
+&nbsp; +
+
+The entities are created using REST API and the association is automatically established between them.
+
+
+==== Camel pipeline
+
+Camel pipeline is an expression of the following business rules
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
The Camel pipeline is an expression of the following business rules:
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+image::camel-component-pipeline.svg[Camel pipeline]
+
+&nbsp; +
+
+
+== Demo
+
+To run the demo successfully you need to have a developer Twitter account with appropriate API keys and secrets.
+
+Got to the application directory and build all components
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
Got to the application directory and build all components:
```

(There should be a ":" after all of the following sentences leading to code)

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+image::camel-component-mailhog.png[MailHog messages]
+
+&nbsp; +
+
+
+== Code walkthrough
+
+To use Debezium Camel component we need to add at least following dependencies into a `pom.xml` file
+
+[source,xml]
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Can you add the "no-indent" (or remove leading indent) and change two 2 spaces?

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-debezium-postgres</artifactId>
+        </dependency>
+    </dependencies>
+----
+
+The pipeline logic itself is defined in https://github.com/debezium/debezium-examples/blob/master/camel-component/qa-camel/src/main/java/io/debezium/examples/camel/pipeline/QaDatabaseUserNotifier.java[QaDatabaseUserNotifier] class.
+
+The main route looks like
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Maybe show the entire method that hosts this?

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+            .endParent();
+
+
+
+    @Converter
+    public static Question questionFromStruct(Struct struct) {                           // (4)
+        return new Question(struct.getInt64("id"), struct.getString("text"), struct.getString("email"));
+    }
+
+    from(ROUTE_MAIL_QUESTION_CREATE)                                                     // (6)
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]: Same as above. Ideall,y the code is a valid Java class.

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+The message store checks if it has the question already stored.
+If not a new aggregate is created and stored otherwise the stored aggregate is updated with new data.
+6. A supplementary route is invoked that formats a mail message and delivers it to the question creator via the SMTP endpoint.
+7. The route part related to the answer message type is very similar (answers are added to question aggregate).
+The main difference is the posting of a Twitter message when the aggregate contains three answers.
+
+
+== Conclusion
+
+Apache Camel integration is an interesting option for system integration scenarios.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
Apache Camel is a very interesting option for implementing system integration scenarios.
```

File: blog/2020-02-19-debezium-camel-integration.adoc [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+When Kafka broker is not necessary to be deployed in the solution it is very easy to deploy a standalone Camel route with Debezium component that enables capturing of change data and execute very complex routing and transformation operations on it and provides the developer with a full arsenal of enterprise integration patterns and more than hundred connectors for different systems that could be included in a complex service orchestration.
```
Comment (gunnarmorling) [inline blog/2020-02-19-debezium-camel-integration.adoc]:
```suggestion
Without the need for any external messaging infrastructure, it is very easy to deploy a standalone Camel route with the Debezium component, enabling the capture of data changes and execution of complex routing and transformation operations on them.
Camel provides the developer with a full arsenal of enterprise integration pattern implementations, as well as more than hundred connectors for different systems that could be included in a complex service orchestration.
```

Also add a sentence linking to the complete example on GitHub once more.

**Comment by jpechane** [comment]
@gunnarmorling All comments should be addressed

**Comment by gunnarmorling** [comment]
@jpechane, on the architecture overview the Twitter icon is scaled disproportionally. Could you update that visual?

**Comment by gunnarmorling** [comment]
Pushed a commit with a few updates.

Still needs the Docker Compose update once we have it.

**Comment by gunnarmorling** [comment]
@jpechane, pushed one more commit. Image looking good now, thanks! One more question: can you add a sentence towards the end touching on the idea of an offset storage based on the embedded infinispan store? With that, and the Compose update, it's good to go from my PoV. Great stuff, thanks a lot!

**Comment by jpechane** [comment]
@gunnarmorling Docker compose used, Inifinispan based offset store mentioned

**Comment by gunnarmorling** [comment]
Thanks, great work! Will publish it it later in the day 🚀

**Pull request merged**